Add new @linked-workspace-participations endpoint.

njohner opened the pull request:
With this PR we add a new endpoint `@add-participations-on-workspace` allowing to set participations on a linked workspace from the Gever dossier.
To avoid having to do multiple requests to the teamraum deployment and to simplify error handling we enable setting a list of participants in the `@participations` endpoint, so that we have a single request from Gever to Teamraum and setting the participations either works for all participants or None.

I'm not sure whether I should add translations of the error messages. These errors should actually not happen in the new UI.

For [CA-3587]

## Checklist

- [x] Changelog entry
- [x] Link to issue (Jira or GitHub) and backlink in issue (Jira)
- API change:
  - [ ] Documentation is updated
  - [ ] API Changelog entry (see [guide](https://4teamwork.atlassian.net/wiki/spaces/4TEAM/pages/451248812/API+Changelog+Guidelines))
  - If breaking:
    - [ ] api-change label added
    - [ ] #delivery channel notified about breaking change
    - [ ] Scrum master is informed
- New translations
  - [ ] All msg-strings are unicode

[CA-3587]: https://4teamwork.atlassian.net/browse/CA-3587?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

tinagerber commented on docs/public/dev-manual/api/workspace/participation.rst:
```diff
@@ -153,7 +153,7 @@ Ein DELETE Request auf die `@id` einer Beteiligung löscht die entsprechnede Bet
 Beteiligungen hinzufügen:
 -------------------------
-In einem selbst verwalteten Teamraum-Ordner (Vererbung wurde unterbrochen) können beteiligungen über einen POST request auf den @participations Endpoint hinzugefügt werden.
+In einem selbst verwalteten Teamraum-Ordner (Vererbung wurde unterbrochen) können beteiligungen über einen POST request auf den @participations Endpoint hinzugefügt werden. Im Body werden entweder die Attributen ``participant`` und ``role`` erwartet oder ein Liste von Beteiligungen im Attribut ``participants``.
```
```suggestion
In einem selbst verwalteten Teamraum-Ordner (Vererbung wurde unterbrochen) können Beteiligungen über einen POST Request auf den @participations Endpoint hinzugefügt werden. Im Body werden entweder die Attribute ``participant`` und ``role`` erwartet oder ein Liste von Beteiligungen im Attribut ``participants``.

tinagerber commented on docs/public/dev-manual/api/workspace/participation.rst:
```diff
@@ -194,9 +194,66 @@ In einem selbst verwalteten Teamraum-Ordner (Vererbung wurde unterbrochen) könn
         "title": "Maria Meier (maria.meier)",
         "id": "maria.meier",
         "is_local": null
+        }
```
The indentation is wrong.

tinagerber commented on docs/public/dev-manual/api/workspace/participation.rst:
```diff
@@ -194,9 +194,66 @@ In einem selbst verwalteten Teamraum-Ordner (Vererbung wurde unterbrochen) könn
+
+   .. sourcecode:: http
+
+       POST /workspaces/workspace-1/folder-1/@participations HTTP/1.1
```
```suggestion
       POST /workspaces/workspace-1/@participations HTTP/1.1

tinagerber commented on opengever/api/tests/test_participation.py:
```diff
@@ -872,6 +872,14 @@ def test_let_a_user_participate(self, browser):
             )
 
+        # Posting with participant and role returns the serialized participant
+        self.assertEquals(
```
assertEqual

tinagerber commented on opengever/api/tests/test_participation.py:
```diff
@@ -872,6 +872,14 @@ def test_let_a_user_participate(self, browser):
+        self.assertEquals(
+            self.workspace_member.id,
+            browser.json.get('participant').get('id'))
+        self.assertEquals(
```
assertEqual

tinagerber commented on opengever/api/tests/test_participation.py:
```diff
@@ -884,6 +892,55 @@ def test_let_a_user_participate(self, browser):
+        self.assertEqual(2, len(browser.json))
+
+        entry = get_entry_by_id(browser.json, self.workspace_member.id)
+        self.assertEquals(
```
assertEqual

tinagerber commented on opengever/api/tests/test_participation.py:
```diff
@@ -884,6 +892,55 @@ def test_let_a_user_participate(self, browser):
+            entry.get('role'))
+
+        entry = get_entry_by_id(browser.json, self.workspace_guest.id)
+        self.assertEquals(
```
assertEqual

tinagerber commented on docs/public/dev-manual/api/linked_workspaces.rst:
```diff
@@ -126,6 +126,64 @@ Eine bestehende Verknüpfung eines Teamraums kann mit dem ``@unlink-workspace``
+    HTTP/1.1 200 OK
+    Content-Type: application/json
+
+    [
```
I'm not so sure we want to return lists in the API. Maybe @lukasgraf  wants to give his opinion?

njohner replied:
yes you are right, I'll return an object instead.

tinagerber reviewed (commented):
Thanks for translating all these error messages 🎉

njohner commented:
Ok @tinagerber can you have another look? I've implemented the changes you pointed out. After discussing with Lukas I've also renamed the endpoint to `linked-workspace-participations`. Endpoint name should not contain the verb, as the verb is the method used. This allows us to maybe implement a GET, DELETE, PATCH in the future...

tinagerber commented on docs/public/dev-manual/api/workspace/participation.rst:
```diff
@@ -194,9 +194,66 @@ In einem selbst verwalteten Teamraum-Ordner (Vererbung wurde unterbrochen) könn
+    HTTP/1.1 200 OK
+    Content-Type: application/json
+
+    [
```
Should return items

njohner replied:
yep sorry

tinagerber commented on docs/public/dev-manual/api/linked_workspaces.rst:
```diff
@@ -126,6 +126,64 @@ Eine bestehende Verknüpfung eines Teamraums kann mit dem ``@unlink-workspace``
+Teilnehmer in einem verknüpften Teamraum setzen
+-----------------------------------------------
+
+Mit dem ``@add-participations-on-workspace`` Endpoint können Teilnehmer auf einem verknüpften Teamraum hinzugefügt werden.
```
The documentation is not up to date.

njohner replied:
Jesus christ I had done that change, but seems I somehow screwed up rebasing 😓
